### ISSUE TITLE
DynamoDB: omit empty fields from requests

### ISSUE THREAD

**Issue by kim** [brendanhay/amazonka#98]
DynamoDB seems to consider fields which are present, but empty, an error:

```haskell
> runAWST env . sendCatch $ createTable "test" (keySchemaElement "id" Hash :| []) (provisionedThroughput 10 10) & ctAttributeDefinitions .~ [attributeDefinition "id" S]

Right (Left (ServiceError "DynamoDB" (Status {statusCode = 400, statusMessage = "Bad Request"}) (JSONError {_jsonType = Just "com.amazon.coral.validate#ValidationException", _jsonCode = Nothing, _jsonMessage = "One or more parameter values were invalid: List of LocalSecondaryIndexes is empty"})))
```

```haskell
> runAWST env . sendCatch $ putItem "test" & piItem .~ (fromList [("id", attributeValue & avS .~ Just "someid")]) & piConditionExpression .~ Just ("id <> :id") & piExpressionAttributeValues .~ (fromList [(":id", attributeValue & avS .~ Just "someid")])

Left (ServiceError "DynamoDB" (Status {statusCode = 400, statusMessage = "Bad Request"}) "JSONError {_jsonType = Just \"com.amazon.coral.validate#ValidationException\", _jsonCode = Nothing, _jsonMessage = \"Can not use both expression and non-expression parameters in the same request: Non-expression parameters: {Expected} Expression parameters: {ConditionExpression}\"}")
```

Since I don't know about other AWS services, I'm not sure what the best solution for this would be. Perhaps wrap all optional fields in `Maybe` (even if they're lists) and omit `Nothing`-valued fields from the JSON?

**Comment by kim** [comment]
I guess it might be necessary to introduce some other type (eg. `Optional`), as `Nothing == null` atm, which might be desired in the general case.

**Comment by brendanhay** [comment]
Right, this seems (at present) only affect the JSON styled services due to how Aeson does the serialisaiton. Neither the XML nor Query serialisers are representing an empty list value in the request over the wire.

Potentially the `Optional` type can be internal in the following fashion:

```haskell
data Optional a
    = Present a
    | Absent

-- | Invalid Iso, should be a Prism but exists for ease of composition
-- with the current 'Lens . Iso' chaining to hide internal types from the user.
_Present :: Monoid a => Iso' (Optional a) a
_Present = iso f Present
  where
    f (Present x) = x
    f Absent      = mempty
```

1. The smart constructor for a data type sets any `Optional` fields to `Absent`:

```haskell

data FooRequest = FooRequest
    { _fooInt  :: Int
    , _fooList :: Optional (List "AttributeValueList" Text)
    }

fooRequest :: Int -> FooRequest
fooRequest p1 = FooRequest
    { _fooInt = p1
    , _fooList = Absent
    }
```

2. The corresponding lens which points to an `Optional` field then sets the value (forever-more) to `Present`:

```haskell
fooList :: Lens' FooRequest [Text]
fooList = lens _fooList (\s a -> s { _fooList = a }) . _Optional . _List 

```
 
So if the request type is created using the smart constructor but the optional field is never touched, it can be ignored by whatever serialisation rules are created for `Optional` for a specific service type.

This has the drawback of once you've used `set/.~ or modify/%~` etc. you can no longer set the value back to `Absent`.

What do you think - too much magic? Should it just be exposed to the user?

**Comment by kim** [comment]
Sounds like a reasonable tradeoff -- outcry from the bidirectional programming community vs. having to deal with another layer of wrapping.

It might be worth considering, however, to always treat "list" typed fields as `NonEmpty` so as to minimise unintentionally empty values.

**Comment by brendanhay** [comment]
I guess since `Maybe (NonEmpty a)` be analogous to `Optional [a]`:

```haskell
Just (1 :| []) == Present [1]
Nothing       == Absent
```

You could expose `Maybe (NonEmpty a)` instead of `[a]` and do away with the setter magic from above.

Is that what you're getting at?

An issue (completely unverified) might be that if you are attempting a `PUT`/update of some field, and are trying to set something to the empty list `[]` (aka deleting that field's value) you have no way of expressing that since `Nothing` would result in no value being sent over the wire. With the `Optional`/setter variant you could simply `.~ []` and the inner field would have the value `Present []`.

**Comment by kim** [comment]
Hm, I think I was over-eagerly generalizing because [ctAttributeDefinitions](http://hackage.haskell.org/package/amazonka-dynamodb-0.3.4/docs/Network-AWS-DynamoDB-CreateTable.html#v:ctAttributeDefinitions) is optional, where it actually should be required (and non-empty). 

Your original proposal does make sense, and I couldn't find an argument for needing to be able to reset some field to `Absent`.

**Comment by brendanhay** [comment]
OK, so if you set `ctAttributeDefinitions` the API call works as expected?

If so I'll confirm and open an upstream issue regarding the definitions, as well as adding the necessary override + regenerating to make that field required + `NonEmpty`.

**Comment by kim** [comment]
> if you set ctAttributeDefinitions the API call works as expected?

No, it's two unrelated issues: 

a) it doesn't work because other optional fields are present in the JSON payload, but empty (first one the endpoint validator is complaining about is `LocalSecondaryIndexes`)

b) `ctAttributeDefinitions` shouldn't be an optional field in the first place. If I read it correctly, the [service definition JSON](https://github.com/brendanhay/amazonka/blob/develop/gen/input/dynamodb/2012-08-10.normal.json#L808) is correct, but the [output](https://github.com/brendanhay/amazonka/blob/develop/gen/output/dynamodb.json#L6768) is not. According to the [docs](http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_AttributeDefinition.html) there should also be length constraints (resulting in a `NonEmpty` in the generated code), but there [are none](https://github.com/brendanhay/amazonka/blob/develop/gen/input/dynamodb/2012-08-10.normal.json#L514). That should possibly be reported upstream.

**Comment by brendanhay** [comment]
Great, thanks for clarifying.

**Comment by brendanhay** [comment]
FYI I haven't forgotten this (or #101), but am subject(ed) to health/study/employment concerns currently.

**Comment by kim** [comment]
Don't worry, I can work around it for a while. Thanks for the update.

**Comment by brendanhay** [comment]
So, short version: I fucked up with assumptions regarding how anything which had some constraints (such as `minLength > 0`) would be added to the containing shape's `required` set. This had two follow on effects, it would be promoted to the smart constructor's parameter list, and would never receive a `TyApp Maybe` application during type solving.

This is despite some fields (prominently in DynamoDB) being `NonEmpty`, but not actually required.
In reality the information was there that it should implicitly be `Maybe (NonEmpty a)`.

This has been fixed with fb1bd5e and should result in no fields ever being promoted to the required parameter list, ensuring fields which are not in the `required` set for a shape always being wrapped with `Maybe`.

I'm not convinced this will 100% solve all cases where the above issue arises. Need to ponder more.

Required parameters that have types that can be defaulted (currently only `[a]` via `mempty`) will be demoted from the smart constructor parameters automatically.
